### PR TITLE
(ACTION): string literal joiner

### LINKED ISSUE #195: Support splitting and joining string literals
According to _Rust Reference: 3.5.1.2.2_ [string literals][string literals] can span multiple lines.

It would be nice to able to split the string literals with `Enter` and join with `Join Lines` action.

### Suggested fixtures

Before splitting:

```rust
fn main() {
    let lit = "Hello, <caret>World";
    println!("{}", lit);
}
```

After splitting:

```rust
fn main() {
    let lit = "Hello, \
               <caret>World";
    println!("{}", lit);
}
```

Before joining:

```rust
fn main() {
    let lit = "He<caret>llo, \
               World";
    println!("{}", lit);
}
```

After joining:

```rust
fn main() {
    let lit = "Hello, <caret>World";
    println!("{}", lit);
}
```

 [string literals]: http://doc.rust-lang.org/reference.html#string-literals

### PR DESCRIPTION
**Pull request by mkaput**
fix #195

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandler.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+        val text = document.charsSequence
+        val leftPsi = file.findElementAt(offsetNear) ?: return CANNOT_JOIN
+        val rightPsi = file.findElementAt(end) ?: return CANNOT_JOIN
+
+        if (leftPsi == rightPsi && leftPsi.elementType in RustTokenElementTypes.STRING_LITERALS) {
+            var start = offsetNear
+
+            // Strip newline escape
+            if (CharSequenceSubSequence(text, 0, start + 1).endsWithUnescapedForwardSlash()) {
```
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandler.kt]: Omg IntelliJ does string copying of entire document here o_O [source](https://upsource.jetbrains.com/idea-ce/file/idea-ce-5418ee4421e14e3eeb4cb10eca6407ee9618379e/plugins/properties/src/com/intellij/lang/properties/PropertiesJoinLinesHandler.java?nav=1162:1200:focused&line=0)

File: src/main/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandler.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+class RustStringLiteralJoinLinesHandler : JoinLinesHandlerDelegate {
+    override fun tryJoinLines(document: Document, file: PsiFile, offsetNear: Int, end: Int): Int {
+        if (file !is RustFile) return CANNOT_JOIN
+
+        val text = document.charsSequence
+        val leftPsi = file.findElementAt(offsetNear) ?: return CANNOT_JOIN
+        val rightPsi = file.findElementAt(end) ?: return CANNOT_JOIN
+
+        if (leftPsi == rightPsi && leftPsi.elementType in RustTokenElementTypes.STRING_LITERALS) {
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandler.kt]: I would have inverted this `if` to save indentation.

File: src/test/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandlerTest.kt [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    fun testNoEscape() = doTest(
+        """
+        fn main() {
+            "Hello<caret>,
+             World"
+        }
+        """,
+        """
+        fn main() {
+            "Hello,<caret> World"
```
Comment (matklad) [inline src/test/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandlerTest.kt]: It does not matter here, but you can use `/*caret*/` now: https://github.com/intellij-rust/intellij-rust/blob/3dd1a3f922f8f59ccfe6b2705781cfb82f2784d4/src/test/kotlin/org/rust/lang/RustTestCaseBase.kt#L91

File: src/test/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandlerTest.kt [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    fun testEscapedButNotEscapedInFactNewlineEscape() = doTest(
+        """
+        fn main() {
+            "He<caret>llo, \\\
+             World"
+        }
+        """,
+        """
+        fn main() {
+            "Hello, \\<caret> World"
```
Comment (matklad) [inline src/test/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandlerTest.kt]: Hm, shouldn't this be `"Hello, \\<caret>\n                                 World"`?

File: src/main/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandler.kt (hunk 2 above)
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandler.kt]: I'm considering implementing joining this:
```rust
"aaa" +
    "bbb"
```

That's why this if has such form

File: src/test/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandlerTest.kt (hunk 4 above)
Comment (mkaput) [inline src/test/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandlerTest.kt]: I don't know. I'm not using this feature on the daily basis, so I may have bad assumptions about it's working.

File: src/main/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandler.kt (hunk 2 above)
Comment (matklad) [inline src/main/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandler.kt]: But do we need this? `let s = "Hello" + "world";` is a compile time error in Rust.

File: src/test/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandlerTest.kt (hunk 4 above)
Comment (matklad) [inline src/test/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandlerTest.kt]: Same here :) I think that adding a `\n` is the correct behavior here, but let's not do this for now and wait for the feedback from users?

Comment (mkaput) [inline src/test/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandlerTest.kt]: We already have fast feedback on such things (looking at you `#` hehe), so this is good idea :)

**Comment by mkaput** [comment]
Looks like it's finised

**Comment by matklad** [comment]
:+1:

**Pull request merged**